subsys/net/lib/http/http_client.c

```diff
@@ -38,6 +38,18 @@ static int sendall(int sock, const void *buf, size_t len)
 			return -errno;
 		}
 
+		if (out_len == 0 && len != 0) {
```

Comment:
Not sure I understand this check, `out_len == 0` means peer closed the connection, so there's no need to `poll()`. Also, `len` is subtracted after this check, so it might be the case that we've already sent all we wanted, but still issue the `poll()` call. 

IMO it'd be better just to call `zsock_poll()` before `zsock_send()`, unconditionally. We should then check the poll result, if the `POLLOUT` was actualy set on a socket (we could timeout). Finally, the error check from `zsock_send()` should verify if the `errno` was set to `EAGAIN` - it should not report an error in such case but just try again.

Reply:
>out_len == 0 means peer closed the connection,   

As far as I understand [the socket send spec](https://man7.org/linux/man-pages/man2/send.2.html) , 0 is the number of bytes we could send  (so none), or timeout indication. This was my assumption. 
Is it different in Zephyr? 

>  it might be the case that we've already sent all we wanted
I call poll only if `out_len `is 0, so subtracting 0 from `len `would not mean that we already send anything

Reply:
Calling zsock_poll _before_ would mean that we call every time, so slowing down sending if we have a lot of small requests, no?

Reply:
Ok, you're right about the first thing, I mixed up return codes with `recv()`, for `send()` 0 doesn't really mean anything else than nothing was send. It is quite unusual though, as a typical return code for a non-blocking socket when it cannot send a data is `-1` with errno set to `EAGAIN`. So if you really want to call `poll()` after initial `send()` I think you should rather check for that, otherwise you'll end up busy looping anyway.

And just FYI, a timeout on a blocking socket, in case `SO_SNDTIMEO` is configured, is also notified by EAGAIN (https://linux.die.net/man/7/socket).

Reply:
https://linux.die.net/man/7/socket
> if no data has been transferred and the timeout has been reached then -1 is returned with errno set to EAGAIN or EWOULDBLOCK,

From https://man7.org/linux/man-pages/man2/send.2.html 
> EAGAIN or EWOULDBLOCK
              The socket is marked nonblocking and the requested
              operation would block.  POSIX.1-2001 allows either error
              to be returned for this case, and does not require these
              constants to have the same value, so a portable
              application should check for both possibilities.

So  we should probably check `errno `(as you suggest) and out_len as well:
`(errno == -EAGAIN || out_len == -EAGAIN)`

Reply:
In no circumstance an error code is returned directly by a socket call - in case of error, it always return -1 and `errno` value is set accordingly (check "Return Value" section in the send doc). So the check before `poll()` should look like (note that errno value is positive):
`(out_len < 0 && errno == EAGAIN)`

Reply:
>POSIX.1-2001 allows either error
to be returned for this case, and does not require these
constants to have the same value

This contradicts the fact that socket always returns -1, but if Zephyr does that all the time and never explicit error code, then it's fine.

As for the check, what I observed during busy looping with non-blocking socket, was that out_len was continuously returning 0, again and again.
So we should include 0 for non-blocking case, no?
```
if (out_len <=0 && errno == EAGAIN) { 
   perform polling
} else if (out_len <0) { // as it was 
return -errno; 
}
```

Reply:
And good point about `EAGAIN` instead `-EAGAIN`, I oversaw it, thanks

Reply:
> out_len was continuously returning 0, again and again.

That's quite surprising, I'd really expect EAGAIN to be returned. But I guess you're right, we need to take 0 into account as well.

> if (out_len <=0 && errno == EAGAIN) { 
>    perform polling
> } else if (out_len <0) { // as it was 
> return -errno; 
> }

If `out_len == 0` we shouldn't check errno, as it shouldn't be set in such case, so I'd say the condition should look like:
```
if (out_len == 0 || (out_len < 0 && errno == EAGAIN) {
    /* poll */
} else if (out_len < 0) {
    return -errno;
}
```

Reply:
Yes, looks great.
This seems to cover all cases - first part of OR for non-blocking and second part of OR for blocking.

Reply:
I'll push correspondent changes.